### PR TITLE
Closes #3: Added singleton factory to migration

### LINKED ISSUE #3: Add support for SingletonFactory in migrations
EIP: https://eips.ethereum.org/EIPS/eip-2470
Example: https://github.com/gnosis/1155-to-20/blob/master/migrations/1_deploy-all.js

### PR DESCRIPTION
**Pull request by rmeissner**
Closes #3

- Extracted the single deployment logic into separate files, maybe we should even extract this in a separate lib so that we can easily reuse it in other projects.

- Use docker (and docker-compose) so that always the same bytecode can be generated (as the contract address depends on this)

- Added test to check the generated address. By running the test you can easily check if the generated bytecode corresponds to the bytecode used for deployment.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,15 @@
+FROM node:12
+
+ENV USER=root
+WORKDIR "/"
+
+COPY package.json ./
```
Comment (Uxio0) [inline Dockerfile:6]: You should be able to do all these copies in the same stage `COPY package.json yarn.lock truffle-config.js networks.json ./`

**Review by Uxio0 — APPROVED** [review]

File: Dockerfile (hunk 1 above)
Comment (cag) [reply Dockerfile]: I actually would go further and remove these `COPY`s, and instead moving the `COPY . .` statement here.

**Review by cag — APPROVED** [review]
You can also remove the `Migrations.sol` file and everything will still work.

File: Dockerfile (hunk 1 above)
Comment (Uxio0) [reply Dockerfile]: @cag If you `COPY` everything nothing is cached, and that way when a single file is changed all the dependencies are fetched again. This way dependencies are just fetched when any of those files `package.json yarn.lock truffle-config.js networks.json` change

**Pull request merged**